themes/default/content/blog/resource-methods-for-pulumi-packages/index.md

```diff
@@ -0,0 +1,609 @@
+        "properties": {
+          "result": {
+            "type": "string"
+          }
+        },
+        "required": ["result"]
+      }
+    }
+  },
+```
+
+Our method has two required arguments: `__self__` and `recipient`. `__self__` is a special input required by all methods that represents the component resource and is typed as such. The method has one result named `result`.
+
+Next, define our `Message` component:
+
+```json
+  "resources": {
+    "example:index:Message": {
+      "isComponent": true,
+      "inputProperties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "requiredInputs": ["message"],
+      "properties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "required": ["message"],
+      "methods": {
+        "getMessage": "example:index:Message/getMessage"
+      }
+    }
+  },
+```
+
+Our component has a single required input/output property: `message`. The method is specified in the `methods` property, which references the method's function definition: `"getMessage": "example:index:Message/getMessage"`.
+
+Altogether our schema looks like:
+
+```json
+{
+  "version": "0.0.1",
+  "name": "example",
+  "functions": {
+    "example:index:Message/getMessage": {
+      "inputs": {
+        "properties": {
+          "__self__": {
+            "$ref": "#/resources/example:index:Message"
+          },
+          "recipient": {
+            "type": "string"
+          }
+        },
+        "required": ["__self__", "recipient"]
+      },
+      "outputs": {
+        "properties": {
+          "result": {
+            "type": "string"
+          }
+        },
+        "required": ["result"]
+      }
+    }
+  },
+  "resources": {
+    "example:index:Message": {
+      "isComponent": true,
+      "inputProperties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "requiredInputs": ["message"],
+      "properties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "required": ["message"],
+      "methods": {
+        "getMessage": "example:index:Message/getMessage"
+      }
+    }
+  },
+  "language": {
+    "csharp": {
+      "packageReferences": {
+        "Pulumi": "3.12"
+      },
+      "liftSingleValueMethodReturns": true
+    },
+    "go": {
+      "liftSingleValueMethodReturns": true
+    },
+    "nodejs": {
+      "devDependencies": {
+        "@types/node": "latest"
+      },
+      "liftSingleValueMethodReturns": true
+    },
+    "python": {
+      "liftSingleValueMethodReturns": true
+    }
+  }
+}
+```
+
+To make it easier for users of our method, we also set `"liftSingleValueMethodReturns": true` for each language, which makes single value methods return the single value directly, rather than wrapping the results in a `Result` type.
+
+### Component Implementation
+
+We can implement the component in TypeScript/JavaScript, Python, or Go, and make it available to any Pulumi language.
```

Comment:
I know we don't have a C# example right now. Feasibly, someone could use C#, right? We probably should adjust the wording here as follows to avoid someone asking why the system isn't available in C#:

```suggestion
We can implement the component and make it available to any Pulumi language. We're going to show some implementation examples in TypeScript/JavaScript, Python, and Go.
```